### DISCUSSION TITLE
Async file io

### DISCUSSION THREAD

**Discussion by 93578237** [RobertCraigie/prisma-client-py#357] [Ideas]
What about using [aiofiles](https://github.com/tinche/aiofiles)?

**Comment by RobertCraigie** [comment]
@q0w What part of this project do you think would benefit from using aiofiles?

**Reply by 93578237** [reply]
http?

**Reply by 93578237** [reply]
I also think, that switching from httpx to aiohttp would improve performance.

**Reply by RobertCraigie** [reply]
As in the internal queries or the binary downloading?

**Reply by RobertCraigie** [reply]
We can't switch from httpx to aiohttp because we need to support both asynchronous and synchronous use cases.

**Reply by 93578237** [reply]
binary

**Reply by 93578237** [reply]
we can use https://github.com/python-trio/unasync

**Reply by RobertCraigie** [reply]
I originally wrote this library to support both `aiohttp` and `requests` but to avoid installing redundant dependencies users needed to specify an extra when they installed the library and this was not a good UX.

**Reply by RobertCraigie** [reply]
That still leads to the redundant dependency problem and I believe that native synchronous / asynchronous solutions will lead to better performance.

**Reply by RobertCraigie** [reply]
Also before switching http libraries to improve performance for binary downloading there are other potential improvements that could be made first such as parallelising requests.

**Reply by 93578237** [reply]
>parallelising requests

using httpx?

**Reply by RobertCraigie** [reply]
Yes, I haven't used it before but that should be possible with any http library. You could use threads / multiprocessing if the library doesn't support it natively.
